### PR TITLE
ArrayBuffer and Binary json encoding

### PR DESCRIPTION
**Pull request by yourWaifu**
I needed this for a project, and so I implemented them for myself. However, I noticed that others asked about it here, #68, so I decided to share them.

Example:

```typescript
socket.on("data", (chunk) => {
  // imagine IPC between a program that sends their chunks with a 4 byte length little endian prefix
  const lengthLength = 4;
  let length = 0;
  for (var i = 0; i < lengthLength; i += 1) {
	  length = (length * 256) + chunk[i];
  }
  
  // create QuickJS ArrayBuffer
  let bufferStart = lengthLength;
  let bufferEnd = length + lengthLength;
  this._contextJS.newArrayBuffer(new Uint8Array(chunk.slice(bufferStart, bufferEnd)))
  .consume((handle) => {
	  // decode data inside the buffer
	  this._contextJS?.decodeBinaryJSON(handle)
		  .consume((handle) => console.log(this._contextJS?.dump(handle)));
  });
```

### CODE REVIEW

**Review by justjake — APPROVED** [review]
Thanks for the contribution! Your changes look good. I have a suggestion for improving the efficiency of copying a ArrayBuffer into QuickJS, and it would also be nice to see some tests.

File: ts/context.ts [hunk 1]
```diff
@@ -932,4 +943,22 @@ export class QuickJSContext implements LowLevelJavascriptVm<QuickJSHandle>, Disp
 
     return this.newError(error)
   }
+
+  /**
+   * Outputs QuickJS Objects in binary form
+   * 
+   * **WARNING**: QuickJS's binary JSON doesn't have a standard so expect it to change between version
+   */
+  encodeBinaryJSON(handle: QuickJSHandle): QuickJSHandle {
```
[the remark this answers is not in the document]
Comment (yourWaifu) [reply ts/context.ts:952]: I mostly have been using decodeBinaryJSON to handle IPC between an app that uses quickJS and one that uses node.js. However, I realized that encodeBinaryJSON isn't all that useful without some sort of getArrayBuffer function. I've been trying to write one, but it seems to not want to work.

**Comment by yourWaifu** [comment]
I added a getArrayBuffer function, but the values outputed by it seem to be incorrect for unknown reasons, so I'll put this into draft until that can be fixed

**Comment by justjake** [comment]
If the array buffer creation bit works well, can we merge that without the JSONB parts?

**Comment by yourWaifu** [comment]
After some experiments, I noticed that NewArrayBufferCopy and NewArrayBuffer creates different outputs when using GetAreayBuffer. I'm guess you are supposed to read the buffet differently but not sure how the code knows the difference and how it reads Array Buffer.

File: ts/context.ts [hunk 2]
```diff
@@ -378,9 +378,9 @@ export class QuickJSContext implements LowLevelJavascriptVm<QuickJSHandle>, Disp
    */
   newArrayBuffer(buffer: ArrayBufferLike): QuickJSHandle {
     const array = new Uint8Array(buffer);
-    const ptr = this.memory
+    const handle = this.memory
       .newHeapBufferPointer(array)
-      .consume((bufferHandle) => this.ffi.QTS_NewArrayBuffer(this.ctx.value, bufferHandle.value.pointer, array.length))
```
Comment (yourWaifu) [inline ts/context.ts]: removing this line seemed to fix the issue. Not sure if it'll cause more issues, but it seems to be fine

File: c/interface.c [hunk 3]
```diff
@@ -319,6 +327,25 @@ JSBorrowedChar *QTS_GetString(JSContext *ctx, JSValueConst *value) {
...
+    return 0;
+  uint8_t *result = malloc(length);
+  if (!result)
+    return result;
+  memcpy(result, buffer, length);
+  return result;
+}
+
+// I don't know how to return two values in C, maybe allocate memory in stack?
+size_t QTS_GetArrayBufferLength(JSContext *ctx, JSValueConst *data) {
```
Comment (justjake) [inline c/interface.c:343]: I think we could add `stackAlloc` and use it to pass in a `*size_t` argument like the `&length` arguments in QuickJS's own API, but it's not document anywhere so I think your current approach is fine.

**Review by justjake — APPROVED** [review]
I would love to see some tests for this stuff. Other than that, it looks ready to merge.

File: c/interface.c (hunk 3 above)
Comment (yourWaifu) [reply c/interface.c:343]: Looking online, the best solutions I could find is to use Emscripten's embind API and a separate cpp file to use the API. However, I'm not familiar with Makefile, so I don't really know how to get it to compile the separate cpp file.

**Comment by yourWaifu** [comment]
It still says awaiting for approval

**Comment by sgrove** [comment]
I also came across a use case where I need to share an ArrayBuffer between host and client, it'd be great to see this merged in!

**Comment by stickmy** [comment]
I has the same scenario, but i went some issues.

**javascript code executed in quickjs**

```js
console.log(new Float32Array([-1, -1, 1, -1, -1, 1, 1, 1]));
```

When i use `vm.getArrayBuffer()` in `console.log` implementation, i got a error:
`Couldn't allocate memory to get ArrayBuffer
    at QuickJSContext.getArrayBuffer`

**Pull request merged**